src/node/miner.cpp

```diff
@@ -193,31 +194,55 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock(const CScript& sc
+            }
+            coinbaseTx.vin[index].scriptSig = CScript() << nHeight << OP_0;
+            // Non-consensus commitment output before finishing coinbase transaction
+            if (commit_scripts && !commit_scripts->empty()) {
+                for (auto commit_script: *commit_scripts) {
+                    coinbaseTx.vout.insert(std::prev(coinbaseTx.vout.end()), CTxOut(policyAsset, 0, commit_script));
+                }
+            }
+            pblock->vtx[index] = MakeTransactionRef(std::move(coinbaseTx));
+            pblocktemplate->vchCoinbaseCommitment = GenerateCoinbaseCommitment(*pblock, pindexPrev, chainparams.GetConsensus());
+            pblocktemplate->vTxFees[index] = fee.second;
+            index++;
         }
-        // 0-value outputs must be unspendable
-        if (coinbaseTx.vout[0].nValue.GetAmount() == 0) {
-            coinbaseTx.vout[0].scriptPubKey = CScript() << OP_RETURN;
+    } else {
+        coinbaseTx.vout.resize(1);
+        coinbaseTx.vout[0].scriptPubKey = scriptPubKeyIn;
+        coinbaseTx.vout[0].nAsset = policyAsset;
+        coinbaseTx.vout[0].nValue = nFees + GetBlockSubsidy(nHeight, chainparams.GetConsensus());
+        if (g_con_elementsmode) {
+            if(chainparams.GetConsensus().subsidy_asset != policyAsset) {
+                // Only claim the subsidy if it's the same as the policy asset.
+                coinbaseTx.vout[0].nValue = nFees;
+            }
+            // 0-value outputs must be unspendable
+            if (coinbaseTx.vout[0].nValue.GetAmount() == 0) {
+                coinbaseTx.vout[0].scriptPubKey = CScript() << OP_RETURN;
+            }
         }
-    }
-    coinbaseTx.vin[0].scriptSig = CScript() << nHeight << OP_0;
-    // Non-consensus commitment output before finishing coinbase transaction
-    if (commit_scripts && !commit_scripts->empty()) {
-        for (auto commit_script: *commit_scripts) {
-            coinbaseTx.vout.insert(std::prev(coinbaseTx.vout.end()), CTxOut(policyAsset, 0, commit_script));
+        coinbaseTx.vin[0].scriptSig = CScript() << nHeight << OP_0;
+        // Non-consensus commitment output before finishing coinbase transaction
+        if (commit_scripts && !commit_scripts->empty()) {
+            for (auto commit_script: *commit_scripts) {
+                coinbaseTx.vout.insert(std::prev(coinbaseTx.vout.end()), CTxOut(policyAsset, 0, commit_script));
+            }
         }
-    }
-    pblock->vtx[0] = MakeTransactionRef(std::move(coinbaseTx));
-    pblocktemplate->vchCoinbaseCommitment = GenerateCoinbaseCommitment(*pblock, pindexPrev, chainparams.GetConsensus());
-    pblocktemplate->vTxFees[0] = -nFees;
+        pblock->vtx[0] = MakeTransactionRef(std::move(coinbaseTx));
+        pblocktemplate->vchCoinbaseCommitment = GenerateCoinbaseCommitment(*pblock, pindexPrev, chainparams.GetConsensus());
+        pblocktemplate->vTxFees[0] = -nFees;
 
+    }
```

Comment:
Lets refactor it to accomodate possibility to have block subsidy payed. In our case we don't but we should add this feature as just multi asset fee feature, and we'll set subsidy to 0 in chain params. And lets try to branch the minimum part of code.

Reply:
Refactored to allow block subsidies with any asset fees: https://github.com/SequentiaSEQ/SEQ-Core-Elements/pull/6/commits/b4566a675ca0c4777e598298442ad70a7f97d11c

Also removed a lot of the code duplication.